### PR TITLE
M3 migration with custom theme

### PR DESCRIPTION
**Pull request by android-dev-lxl**
Added app code 
Added Launcher icon
UI feedback

### CODE REVIEW

File: app/src/main/java/com/example/unscramble/ui/GameScreen.kt [hunk 1]
```diff
@@ -0,0 +1,226 @@
...
+            TextButton(onClick = onPlayAgain) {
+                Text(text = stringResource(R.string.play_again))
+            }
+        }
+    )
+}
+
+@Preview(showBackground = true)
+@Composable
+fun GameScreenPreview() {
```
Comment (kkuan2011) [inline app/src/main/java/com/example/unscramble/ui/GameScreen.kt:222]: should this be in the MainActivity file? so we don't need to import Unscramble theme here?

File: app/src/main/java/com/example/unscramble/ui/theme/Color.kt [hunk 2]
```diff
@@ -15,13 +15,69 @@
  */
 
 package com.example.unscramble.ui.theme
-
```
Comment (kkuan2011) [inline app/src/main/java/com/example/unscramble/ui/theme/Color.kt:18]: I think we can keep the new line here

File: app/src/main/java/com/example/unscramble/ui/theme/Color.kt [hunk 3]
```diff
@@ -15,13 +15,69 @@
...
+val md_theme_dark_surfaceTint = Color(0xFFBAC3FF)
+val md_theme_dark_outlineVariant = Color(0xFF46464F)
+val md_theme_dark_scrim = Color(0xFF000000)
+
 
-val Purple40 = Color(0xFF6650a4)
-val PurpleGrey40 = Color(0xFF625b71)
-val Pink40 = Color(0xFF7D5260)
\ No newline at end of file
+val seed = Color(0xFF3F51B5)
```
Comment (kkuan2011) [inline app/src/main/java/com/example/unscramble/ui/theme/Color.kt:83]: what's seed? is that a color provided by the template?

File: app/src/main/java/com/example/unscramble/ui/theme/Theme.kt [hunk 4]
```diff
@@ -20,34 +20,88 @@ import android.app.Activity
...
+    scrim = md_theme_dark_scrim,
 )
 
+
 @Composable
 fun UnscrambleTheme(
     darkTheme: Boolean = isSystemInDarkTheme(),
     // Dynamic color is available on Android 12+
-    dynamicColor: Boolean = true,
+    dynamicColor: Boolean = false,
```
Comment (kkuan2011) [inline app/src/main/java/com/example/unscramble/ui/theme/Theme.kt:104]: Meghan and I talked about adding a comment here so students understand that it's only turned off for learning purposes so they can follow the codelab steps

File: app/src/main/java/com/example/unscramble/ui/theme/Theme.kt [hunk 5]
```diff
@@ -73,4 +130,22 @@ fun UnscrambleTheme(
         typography = Typography,
         content = content
     )
-}
\ No newline at end of file
+}
+
+
+//@Composable
+//fun UnscrambleTheme(
```
Comment (kkuan2011) [inline app/src/main/java/com/example/unscramble/ui/theme/Theme.kt:137]: remove unused code?

File: app/src/main/res/drawable/ic_launcher_foreground.xml [hunk 6]
```diff
@@ -0,0 +1,29 @@
+<vector xmlns:android="http://schemas.android.com/apk/res/android"
```
Comment (kkuan2011) [inline app/src/main/res/drawable/ic_launcher_foreground.xml:1]: add license header to this and double check other files too

File: app/src/main/res/mipmap-anydpi-v26/ic_launcher.xml [hunk 7]
```diff
@@ -16,7 +16,6 @@
   -->
 
 <adaptive-icon xmlns:android="http://schemas.android.com/apk/res/android">
-    <background android:drawable="@drawable/ic_launcher_background" />
-    <foreground android:drawable="@drawable/ic_launcher_foreground" />
-    <monochrome android:drawable="@drawable/ic_launcher_foreground" />
```
Comment (kkuan2011) [inline app/src/main/res/mipmap-anydpi-v26/ic_launcher.xml:21]: do we need this monochrome line in or no?

File: app/src/main/res/values/strings.xml [hunk 8]
```diff
@@ -1,12 +1,10 @@
 <!--
-  ~ Copyright (C) 2023 The Android Open Source Project
+  ~ Copyright (c)2023 The Android Open Source Project
```
Comment (kkuan2011) [inline app/src/main/res/values/strings.xml:2]: nit: is there supposed to be a space after the (C)?

File: app/src/main/res/values/strings.xml [hunk 9]
```diff
@@ -1,12 +1,10 @@
...
   ~
   ~ Licensed under the Apache License, Version 2.0 (the "License");
   ~ you may not use this file except in compliance with the License.
   ~ You may obtain a copy of the License at
   ~
-  ~     https://www.apache.org/licenses/LICENSE-2.0
```
Comment (kkuan2011) [inline app/src/main/res/values/strings.xml:8]: do we need this line?

**Review by kkuan2011 — CHANGES_REQUESTED** [review]
I haven't done an in-depth review of the UI code yet, but can you include a screenshot (in the chat or here)? Thanks!

File: app/src/main/java/com/example/unscramble/MainActivity.kt [hunk 10]
```diff
@@ -38,6 +36,7 @@ class MainActivity : ComponentActivity() {
                     modifier = Modifier.fillMaxSize(),
                     color = MaterialTheme.colorScheme.background
                 ) {
+                    GameScreen()
                 }
             }
         }
```
[the remark this answers is not in the document]
Comment (android-dev-lxl) [reply app/src/main/java/com/example/unscramble/MainActivity.kt:42]: In template code and **JetSurvey** we don't have preview composable in `MainActivity`.
That is not the case with the **Reply** app, it has the preview code in `MainActivity`.
As the composable is in a separate file, I prefer to leave it there. That way students can take advantage of live edit.

File: app/src/main/java/com/example/unscramble/ui/GameScreen.kt (hunk 1 above)
Comment (android-dev-lxl) [reply app/src/main/java/com/example/unscramble/ui/GameScreen.kt:222]: Answered above

File: app/src/main/java/com/example/unscramble/ui/theme/Color.kt (hunk 3 above)
Comment (android-dev-lxl) [reply app/src/main/java/com/example/unscramble/ui/theme/Color.kt:83]: I used the primary color as seed color in the Material theme builder and downloaded the files.

**Comment by android-dev-lxl** [comment]
> I haven't done an in-depth review of the UI code yet, but can you include a screenshot (in the chat or here)? Thanks!

image

**Comment by android-dev-lxl** [comment]
> I haven't done an in-depth review of the UI code yet, but can you include a screenshot (in the chat or here)? Thanks!

image

File: app/build.gradle [hunk 11]
```diff
@@ -32,6 +32,7 @@ android {
     }
     kotlinOptions {
         jvmTarget = '1.8'
+        freeCompilerArgs += "-opt-in=androidx.compose.material3.ExperimentalMaterial3Api"
```
Comment (yenerm) [inline app/build.gradle:35]: missing license

File: app/src/main/java/com/example/unscramble/ui/GameScreen.kt [hunk 12]
```diff
@@ -0,0 +1,226 @@
...
+            }
+        }
+        if (gameUiState.isGameOver) {
+            FinalScoreDialog(
+                score = gameUiState.score,
+                onPlayAgain = { gameViewModel.resetGame() }
+            )
+        }
+    }
+
```
Comment (yenerm) [inline app/src/main/java/com/example/unscramble/ui/GameScreen.kt:113]: remove empty line

**Review by yenerm — CHANGES_REQUESTED** [review]
add licenses on .gradle files

File: app/src/main/res/drawable/ic_star.xml [hunk 13]
```diff
@@ -0,0 +1,5 @@
+<vector android:height="24dp" android:tint="#00105C"
```
Comment (kkuan2011) [inline app/src/main/res/drawable/ic_star.xml:1]: license header

File: app/src/main/java/com/example/unscramble/ui/GameScreen.kt [hunk 14]
```diff
@@ -0,0 +1,272 @@
...
+                .padding(16.dp),
+            verticalArrangement = Arrangement.spacedBy(16.dp),
+            horizontalAlignment = Alignment.CenterHorizontally
+        ) {
+
+            Button(
+                modifier = modifier
+                    .height(65.dp)
+                    .fillMaxWidth()
+                    .padding(top = 16.dp),
```
Comment (kkuan2011) [inline app/src/main/java/com/example/unscramble/ui/GameScreen.kt:104]: If you add more top and bottom padding to the button, can that increase the height of the button and then you can remove the hardcoded height?

**Review by kkuan2011 — COMMENTED** [review]
I didn't get to do an in-depth review of the UI code yet, but I did want to close out my previous comment

File: app/src/main/java/com/example/unscramble/ui/GameScreen.kt (hunk 14 above)
Comment (android-dev-lxl) [reply app/src/main/java/com/example/unscramble/ui/GameScreen.kt:104]: Padding adds space between composables in Compose. 
Height actually increases the size of the composable.

**Review by yenerm — APPROVED** [review]

**Review by yenerm — APPROVED** [review]

File: app/src/main/java/com/example/unscramble/ui/GameScreen.kt [hunk 15]
```diff
@@ -0,0 +1,261 @@
...
+import androidx.compose.material3.MaterialTheme.typography
+import androidx.compose.material3.TextFieldDefaults
+import androidx.compose.ui.draw.clip
+import androidx.compose.ui.res.painterResource
+import androidx.compose.ui.text.font.FontWeight
+import androidx.compose.ui.text.style.TextAlign
+
+@Composable
+fun GameScreen(
+    modifier: Modifier = Modifier,
```
Comment (MagicalMeghan) [inline app/src/main/java/com/example/unscramble/ui/GameScreen.kt:66]: modifier doesn't need to be passed into the main composable

File: app/src/main/java/com/example/unscramble/ui/GameScreen.kt [hunk 16]
```diff
@@ -0,0 +1,261 @@
...
+            wordCount = gameUiState.currentWordCount,
+            userGuess = gameViewModel.userGuess,
+            onKeyboardDone = { gameViewModel.checkUserGuess() },
+            currentScrambledWord = gameUiState.currentScrambledWord,
+            isGuessWrong = gameUiState.isGuessedWordWrong
+        )
+        Column(
+            modifier = modifier
+                .fillMaxWidth()
+                .padding(16.dp),
```
Comment (MagicalMeghan) [inline app/src/main/java/com/example/unscramble/ui/GameScreen.kt:95]: add 16.dp to dimens.xml - along with any other repeating dimensions

Comment (MagicalMeghan) [reply app/src/main/java/com/example/unscramble/ui/GameScreen.kt:95]: add ass padding_medium

File: app/src/main/java/com/example/unscramble/ui/GameScreen.kt [hunk 17]
```diff
@@ -0,0 +1,261 @@
...
+    wordCount: Int,
+    isGuessWrong: Boolean,
+    userGuess: String,
+    onUserGuessChanged: (String) -> Unit,
+    onKeyboardDone: () -> Unit,
+    modifier: Modifier = Modifier
+) {
+    Card(
+        modifier = modifier
+            .fillMaxWidth()
```
Comment (MagicalMeghan) [inline app/src/main/java/com/example/unscramble/ui/GameScreen.kt:161]: have the parent composable pass in the modifier with fillMaxWidth(), wrapContentHeight(), and padding()

File: app/src/main/java/com/example/unscramble/ui/GameScreen.kt [hunk 18]
```diff
@@ -0,0 +1,261 @@
...
+            .wrapContentHeight()
+            .padding(16.dp),
+        elevation = CardDefaults.cardElevation(
```
Comment (MagicalMeghan) [inline app/src/main/java/com/example/unscramble/ui/GameScreen.kt:164]: nit - possibly use an elevated card here

Comment (android-dev-lxl) [reply app/src/main/java/com/example/unscramble/ui/GameScreen.kt:164]: Discussed this over chat

File: app/src/main/java/com/example/unscramble/ui/GameScreen.kt [hunk 19]
```diff
@@ -0,0 +1,261 @@
...
+                onPlayAgain = { gameViewModel.resetGame() }
+            )
+        }
+    }
+}
+
+@Composable
+fun GameStatus(score: Int, modifier: Modifier = Modifier) {
+    Card(
+        modifier = modifier.padding(20.dp)
```
Comment (MagicalMeghan) [inline app/src/main/java/com/example/unscramble/ui/GameScreen.kt:140]: pass this in from the parent

File: app/src/main/java/com/example/unscramble/ui/GameScreen.kt [hunk 20]
```diff
@@ -0,0 +1,261 @@
...
+            modifier = Modifier.padding(mediumPadding)
+        ) {
+            Text(
+                modifier = Modifier
+                    .clip(shapes.medium)
+                    .background(colorScheme.surfaceTint)
+                    .padding(horizontal = 10.dp, vertical = 4.dp)
+                    .align(alignment = Alignment.End),
+                text = stringResource(R.string.word_count, wordCount),
+                fontSize = 18.sp,
```
Comment (MagicalMeghan) [inline app/src/main/java/com/example/unscramble/ui/GameScreen.kt:180]: specify in Type.kt file

File: app/src/main/java/com/example/unscramble/ui/GameScreen.kt [hunk 21]
```diff
@@ -0,0 +1,261 @@
...
+                color = colorScheme.onPrimary
+            )
+            Text(
+                text = currentScrambledWord,
+                fontSize = 45.sp,
```
Comment (MagicalMeghan) [inline app/src/main/java/com/example/unscramble/ui/GameScreen.kt:185]: specify in Type.kt file

**Review by yenerm — APPROVED** [review]

File: app/src/main/java/com/example/unscramble/ui/GameScreen.kt [hunk 22]
```diff
@@ -137,13 +129,12 @@ fun GameScreen(gameViewModel: GameViewModel = viewModel()) {
...
-        modifier = modifier.padding(20.dp)
+        modifier = modifier
     ) {
         Text(
             text = stringResource(R.string.score, score),
-            fontSize = 28.sp,
-            modifier = modifier.padding(8.dp),
-            fontWeight = FontWeight.Bold
+            style = typography.headlineMedium,
+            modifier = modifier.padding(8.dp)
```
Comment (MagicalMeghan) [inline app/src/main/java/com/example/unscramble/ui/GameScreen.kt:137]: Modifier

Comment (MagicalMeghan) [reply app/src/main/java/com/example/unscramble/ui/GameScreen.kt:137]: just this then good to go!

**Pull request merged**